### PR TITLE
Caching implementation

### PR DESCRIPTION
**Pull request by emmary2020**
Cached proxy pattern implemented using https://skrift.io/articles/archive/a-perspective-on-caching-in-umbraco/ as a guide. Installed using DangEasy.Interfaces.Caching &
using DangEasy.Caching.MemoryCache packages to assist.

### CODE REVIEW

File: src/Our.Umbraco.Gandalf/Core/Controllers/Backoffice/AllowedApiController.cs [hunk 1]
```diff
@@ -15,9 +16,9 @@ public class AllowedIpApiController : UmbracoAuthorizedApiController
     {
         private IAllowedIpService _allowedIpService;
 
-        public AllowedIpApiController(IAllowedIpService allowedIpService)
+        public AllowedIpApiController(AllowedIpService allowedIpService)
         {
```
Comment (anthonydang-ry) [inline src/Our.Umbraco.Gandalf/Core/Controllers/Backoffice/AllowedApiController.cs]: This needs to be IAllowedIpService

File: src/Our.Umbraco.Gandalf/Core/Services/CachedProxies/AllowedIpServiceCachedProxy.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
+﻿using System.Collections.Generic;
+using Our.Umbraco.Gandalf.Core.Models.DTOs;
+using Our.Umbraco.OpenKeyValue.Core.Models.Pocos;
+using System.Linq;
+using Umbraco.Core.Cache;
+using DangEasy.Caching.MemoryCache;
```
Comment (anthonydang-ry) [inline src/Our.Umbraco.Gandalf/Core/Services/CachedProxies/AllowedIpServiceCachedProxy.cs]: Remove DangEasy.Caching as we want to use the Umbraco inbuilt cache

File: src/Our.Umbraco.Gandalf/Core/Services/CachedProxies/AllowedIpServiceCachedProxy.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+namespace Our.Umbraco.Gandalf.Core.Services.CachedProxies
+{
+	public class AllowedIpServiceCachedProxy : IAllowedIpService
+	{
+
+		private readonly ICache _cache;
+		private readonly IAllowedIpService _allowedIpService;
+		private readonly IAppCache _runtimeCache;
+
+		public AllowedIpServiceCachedProxy(AllowedIpService allowedIpService, ICache cache, AppCaches appCaches)
```
Comment (anthonydang-ry) [inline src/Our.Umbraco.Gandalf/Core/Services/CachedProxies/AllowedIpServiceCachedProxy.cs]: Remove ICache cache as we want to use appCaches

File: src/Our.Umbraco.Gandalf/Core/StartUp/IpLockerComposer.cs [hunk 4]
```diff
@@ -2,18 +2,39 @@
...
 namespace Our.Umbraco.Gandalf.Core.StartUp
 {
 	public class GandalfComposer : IUserComposer
 	{
+
 		public void Compose(Composition composition)
 		{
-			composition.Register(typeof(IAllowedIpService), typeof(AllowedIpService));
+			composition.Register(typeof(ICache), typeof(Cache));
+
```
Comment (anthonydang-ry) [inline src/Our.Umbraco.Gandalf/Core/StartUp/IpLockerComposer.cs]: This package should always use caching IMO. We don't really want people to turn this off. However we need to turn it off for our own testing.

Perhaps if the appsetting does NOT exist, then assume cache is turned on.

If you WANT to turn it off, then add the app setting = false

File: src/Our.Umbraco.Gandalf/packages.config [hunk 5]
```diff
@@ -4,6 +4,8 @@
   <package id="ClientDependency" version="1.9.7" targetFramework="net472" />
   <package id="ClientDependency-Mvc5" version="1.8.0.0" targetFramework="net452" />
   <package id="CSharpTest.Net.Collections" version="14.906.1403.1082" targetFramework="net472" />
+  <package id="DangEasy.Caching.MemoryCache" version="2.1.0" targetFramework="net472" />
```
Comment (anthonydang-ry) [inline src/Our.Umbraco.Gandalf/packages.config]: Remove DangEasy.Caching & DangEasy.Interfaces

File: src/Our.Umbraco.Gandalf/Core/Services/CachedProxies/AllowedIpServiceCachedProxy.cs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+		public AllowedIpServiceCachedProxy(AllowedIpService allowedIpService, AppCaches appCaches)
+		{
+			_allowedIpService = allowedIpService;
+			_runtimeCache = appCaches.RuntimeCache;
+		}
+	
+		public AllowedIpDto Create(string ipAddress, string notes)
+		{
+			_runtimeCache.ClearByRegex($"{typeof(AllowedIpServiceCachedProxy)}_(.*)");
+			return this._allowedIpService.Create(ipAddress, notes);
```
Comment (anthonydang-ry) [inline src/Our.Umbraco.Gandalf/Core/Services/CachedProxies/AllowedIpServiceCachedProxy.cs]: You will need to swap these, as there can be concurrency issues when multiple backoffice users refresh the backoffice.

var result = this._allowedIpService.Create(ipAddress, notes);
_runtimeCache.ClearByRegex($"{typeof(AllowedIpServiceCachedProxy)}_(.*)");

return result;

Comment (emmary2020) [reply src/Our.Umbraco.Gandalf/Core/Services/CachedProxies/AllowedIpServiceCachedProxy.cs]: 👍

File: src/Our.Umbraco.Gandalf/Core/Services/CachedProxies/AllowedIpServiceCachedProxy.cs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+		public AllowedIpDto GetById(int id)
+		{
+			return _allowedIpService.GetById(id);
+		}
+
+		public AllowedIpDto GetByIpAddress(string ipAddress)
+		{
+			var cacheKey = $"{typeof(AllowedIpServiceCachedProxy)}";
+			return GetAll().FirstOrDefault(a =>a.IpAddress.Equals(ipAddress));
+		}
```
Comment (anthonydang-ry) [inline src/Our.Umbraco.Gandalf/Core/Services/CachedProxies/AllowedIpServiceCachedProxy.cs]: GetByIpAddress isnt using the cache.

return _runtimeCache.GetCacheItem(cacheKey, () => GetByIpAddress .GetAll(cacheKey));

File: src/Our.Umbraco.Gandalf/Core/Services/CachedProxies/AllowedIpServiceCachedProxy.cs [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+		public KeyValueDto GetStatus()
+		{
+			return this._allowedIpService.GetStatus();
+		}
+
+		public AllowedIpDto Update(AllowedIpDto poco)
+		{
+			_runtimeCache.ClearByRegex($"{typeof(AllowedIpServiceCachedProxy)}_(.*)");
+			return this._allowedIpService.Update(poco);
+		}
```
Comment (anthonydang-ry) [inline src/Our.Umbraco.Gandalf/Core/Services/CachedProxies/AllowedIpServiceCachedProxy.cs]: Swap these lines - same issue as create

File: src/Our.Umbraco.Gandalf/Core/Services/CachedProxies/AllowedIpServiceCachedProxy.cs [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+		public KeyValueDto UpdateAppStatus(string value)
+		{
+			_runtimeCache.ClearByRegex($"{typeof(AllowedIpServiceCachedProxy)}_(.*)");
+			return this._allowedIpService.UpdateAppStatus(value);
+		}
```
Comment (anthonydang-ry) [inline src/Our.Umbraco.Gandalf/Core/Services/CachedProxies/AllowedIpServiceCachedProxy.cs]: Swap these lines - same issue as create

File: src/Our.Umbraco.Gandalf/Our.Umbraco.Gandalf.csproj [hunk 10]
```diff
@@ -48,6 +48,12 @@
     <Reference Include="CSharpTest.Net.Collections, Version=14.906.1403.1082, Culture=neutral, PublicKeyToken=06aee00cce822474, processorArchitecture=MSIL">
       <HintPath>..\packages\CSharpTest.Net.Collections.14.906.1403.1082\lib\net40\CSharpTest.Net.Collections.dll</HintPath>
     </Reference>
+    <Reference Include="DangEasy.Caching.MemoryCache, Version=2.1.0.0, Culture=neutral, processorArchitecture=MSIL">
+      <HintPath>..\packages\DangEasy.Caching.MemoryCache.2.1.0\lib\net472\DangEasy.Caching.MemoryCache.dll</HintPath>
+    </Reference>
+    <Reference Include="DangEasy.Interfaces.Caching, Version=1.0.0.0, Culture=neutral, processorArchitecture=MSIL">
+      <HintPath>..\packages\DangEasy.Interfaces.Caching.2.1.0\lib\net472\DangEasy.Interfaces.Caching.dll</HintPath>
+    </Reference>
     <Reference Include="Examine, Version=1.0.0.0, Culture=neutral, processorArchitecture=MSIL">
```
Comment (anthonydang-ry) [inline src/Our.Umbraco.Gandalf/Our.Umbraco.Gandalf.csproj]: Remove DangEasy

File: src/Our.Umbraco.Gandalf/Core/Services/CachedProxies/AllowedIpServiceCachedProxy.cs (hunk 7 above)
Comment (emmary2020) [reply src/Our.Umbraco.Gandalf/Core/Services/CachedProxies/AllowedIpServiceCachedProxy.cs]: Like this?

`return _runtimeCache.GetCacheItem(cacheKey, () => GetByIpAddress(ipAddress));`

**Pull request merged**